### ISSUE TITLE
[BUG] Node Disk Tags Go Missing At Restart Of Node

### ISSUE THREAD

**Issue by irishgordo** [harvester/harvester#2832]
**Describe the bug**
Friday, Sept 23rd, I created on the Dell PowerEdge R720 a StorageClass that targeted a [3TB Hitatchi 7.2k RPM SATA 6Gb/s Drive](https://www.newegg.com/hitachi-gst-ultrastar-7k3000-hus723030als640-3tb/p/N82E16822145628).
Today, upon rebooting the node, Monday, Sept 26th - I was noticing an issue with the VM building it's PVC, as it was stuck in `pending` upon looking at the logs with K9s it mentioned that there is no StorageClass by the tag `spinning` which is what I had built on the 9/23.
Upon looking at Longhorn I could tell as well as via the UI of the Node Configuration, that indeed the disk tag was empty.
When I added the disk tag back to the disk, re-deployed the VM, everything did work.
But the disk tag was never removed, it just disappeared / wasn't present at reboot of the node.

**To Reproduce**
- TBD, will update

**Expected behavior**
The disk tags not to disappear at reboot of node or when node comes back online after being powered off for some time.

**Supporting Screenshots**
- Friday, Sept 23rd, Disk Tag set up on IO media:
![disk-with-storage-class](https://user-images.githubusercontent.com/5370752/192404154-41706527-7d6c-410a-866a-f834bc6c5f76.png)

**Today Screenshots**
![Screenshot from 2022-09-26 17-21-24](https://user-images.githubusercontent.com/5370752/192404228-3f33b301-4648-4494-a3cf-05fe52832e50.png)
![Screenshot from 2022-09-26 17-10-11](https://user-images.githubusercontent.com/5370752/192404230-e13df9e3-0dd3-4cd6-b88f-945304565a6c.png)
![Screenshot from 2022-09-26 17-09-12](https://user-images.githubusercontent.com/5370752/192404231-a72eb676-5dac-417d-80b6-aaa6562ea2c5.png)
![Screenshot from 2022-09-26 16-58-57](https://user-images.githubusercontent.com/5370752/192404232-c81f8ce3-ecb8-48f1-8ceb-7b9ee156b9a7.png)
![Screenshot from 2022-09-26 16-58-28](https://user-images.githubusercontent.com/5370752/192404233-1d1b02d9-4218-4237-9383-ebc8f04f7f09.png)
![Screenshot from 2022-09-26 16-57-26](https://user-images.githubusercontent.com/5370752/192404237-146065d1-2561-4131-92ce-302426eb687f.png)
![disk-is-present](https://user-images.githubusercontent.com/5370752/192404240-dc6cb489-680f-44f1-8b21-0862c2ef81f9.png)
![hung-on-storage](https://user-images.githubusercontent.com/5370752/192404241-26644e9c-83c8-4d70-9c04-7a10bcd35d0e.png)
![spinning-disk-issues-with-vm](https://user-images.githubusercontent.com/5370752/192404242-d92cc127-3999-4747-8bd4-206f15c266f5.png)

**Support bundle**
[supportbundle_22241fc2-4f6e-4105-9483-178b97ed77b8_2022-09-27T00-02-32Z.zip](https://github.com/harvester/harvester/files/9650471/supportbundle_22241fc2-4f6e-4105-9483-178b97ed77b8_2022-09-27T00-02-32Z.zip)

**Environment**
 - Harvester ISO version: v1.1.0-rc1
 - Underlying Infrastructure: Dell PowerEdge R720

**Additional Info**
```
suse-workstation-team-harvester➜  harvester : master ✘ :✹✭ ᐅ  kubectl --kubeconfig ~/.kube/dell-poweredge.yaml get bundle -A 
NAMESPACE     NAME                                          BUNDLEDEPLOYMENTS-READY   STATUS
fleet-local   fleet-agent-local                             1/1                       
fleet-local   local-managed-system-agent                    1/1                       
fleet-local   mcc-harvester                                 0/1                       Modified(1) [Cluster fleet-local/local]; customresourcedefinition.apiextensions.k8s.io engineimages.longhorn.io modified {"status":{"acceptedNames":{"kind":"","plural":""},"conditions":[],"storedVersions":[]}}; customresourcedefinition.apiextensions.k8s.io nodes.longhorn.io modified {"status":{"acceptedNames":{"kind":"","plural":""},"conditions":[],"storedVersions":[]}}; customresourcedefinition.apiextensions.k8s.io volumes.longhorn.io modified {"status":{"acceptedNames":{"kind":"","plural":""},"conditions":[],"storedVersions":[]}}
fleet-local   mcc-harvester-crd                             1/1                       
fleet-local   mcc-local-managed-system-upgrade-controller   1/1                       
fleet-local   mcc-rancher-logging                           1/1                       
fleet-local   mcc-rancher-logging-crd                       1/1                       
fleet-local   mcc-rancher-monitoring                        1/1                       
fleet-local   mcc-rancher-monitoring-crd                    1/1    
```

**Comment by irishgordo** [comment]
More screenshots of the creation can be seen in this:
https://github.com/harvester/harvester/issues/2826
(as that was from the day where the disk was first introduced and tagged)

**Comment by futuretea** [comment]
The disk tag on one of my environments is also missing. I did not remove it.
but I cannot find the reproduce steps

**Comment by irishgordo** [comment]
@futuretea 
Powering on my R720 today, I notice the same thing.

![Screenshot from 2022-09-27 10-19-53](https://user-images.githubusercontent.com/5370752/192599083-627dd1af-2462-4066-91aa-80722ad67483.png)

With the GET https://192.168.1.166/v1/harvester/harvesterhci.io.blockdevices - the Response back is:
```
{
  "type": "collection",
  "links": {
    "self": "https://192.168.1.166/v1/harvester/harvesterhci.io.blockdevices"
  },
  "createTypes": {
    "harvesterhci.io.blockdevice": "https://192.168.1.166/v1/harvester/harvesterhci.io.blockdevices"
  },
  "actions": {},
  "resourceType": "harvesterhci.io.blockdevice",
  "revision": "1116004",
  "data": [
    {
      "id": "longhorn-system/dbaa114c4d3cb0f53a83ba607a3a22ef",
      "type": "harvesterhci.io.blockdevice",
      "links": {
        "remove": "https://192.168.1.166/v1/harvester/harvesterhci.io.blockdevices/longhorn-system/dbaa114c4d3cb0f53a83ba607a3a22ef",
        "self": "https://192.168.1.166/v1/harvester/harvesterhci.io.blockdevices/longhorn-system/dbaa114c4d3cb0f53a83ba607a3a22ef",
        "update": "https://192.168.1.166/v1/harvester/harvesterhci.io.blockdevices/longhorn-system/dbaa114c4d3cb0f53a83ba607a3a22ef",
        "view": "https://192.168.1.166/apis/harvesterhci.io/v1beta1/namespaces/longhorn-system/blockdevices/dbaa114c4d3cb0f53a83ba607a3a22ef"
      },
      "apiVersion": "harvesterhci.io/v1beta1",
      "kind": "BlockDevice",
      "metadata": {
        "creationTimestamp": "2022-09-23T21:30:00Z",
        "fields": [
          "dbaa114c4d3cb0f53a83ba607a3a22ef",
          "disk",
          "/dev/sda",
          "/var/lib/harvester/extra-disks/dbaa114c4d3cb0f53a83ba607a3a22ef",
          "dell-poweredge-r720",
          "Active",
          "3d19h"
        ],
        "finalizers": [
          "wrangler.cattle.io/harvester-block-device-handler"
        ],
        "generation": 7,
        "labels": {
          "kubernetes.io/hostname": "dell-poweredge-r720",
          "ndm.harvesterhci.io/device-type": "disk"
        },
        "managedFields": [
          {
            "apiVersion": "harvesterhci.io/v1beta1",
            "fieldsType": "FieldsV1",
            "fieldsV1": {
              "f:spec": {
                "f:fileSystem": {
                  "f:forceFormatted": {},
                  "f:provisioned": {}
                }
              }
            },
            "manager": "harvester",
            "operation": "Update",
            "time": "2022-09-23T21:42:11Z"
          },
          {
            "apiVersion": "harvesterhci.io/v1beta1",
            "fieldsType": "FieldsV1",
            "fieldsV1": {
              "f:metadata": {
                "f:finalizers": {
                  ".": {},
                  "v:\"wrangler.cattle.io/harvester-block-device-handler\"": {}
                },
                "f:labels": {
                  ".": {},
                  "f:kubernetes.io/hostname": {},
                  "f:ndm.harvesterhci.io/device-type": {}
                }
              },
              "f:spec": {
                ".": {},
                "f:devPath": {},
                "f:fileSystem": {
                  ".": {},
                  "f:mountPoint": {}
                },
                "f:nodeName": {}
              },
              "f:status": {
                ".": {},
                "f:conditions": {},
                "f:deviceStatus": {
                  ".": {},
                  "f:capacity": {
                    ".": {},
                    "f:physicalBlockSizeBytes": {},
                    "f:sizeBytes": {}
                  },
                  "f:details": {
                    ".": {},
                    "f:busPath": {},
                    "f:deviceType": {},
                    "f:driveType": {},
                    "f:model": {},
                    "f:serialNumber": {},
                    "f:storageController": {},
                    "f:uuid": {},
                    "f:vendor": {},
                    "f:wwn": {}
                  },
                  "f:devPath": {},
                  "f:fileSystem": {
                    ".": {},
                    "f:LastFormattedAt": {},
                    "f:mountPoint": {},
                    "f:type": {}
                  },
                  "f:partitioned": {}
                },
                "f:provisionPhase": {},
                "f:state": {}
              }
            },
            "manager": "node-disk-manager",
            "operation": "Update",
            "time": "2022-09-23T21:42:21Z"
          }
        ],
        "name": "dbaa114c4d3cb0f53a83ba607a3a22ef",
        "namespace": "longhorn-system",
        "relationships": null,
        "resourceVersion": "406060",
        "state": {
          "error": false,
          "message": "Resource is current",
          "name": "active",
          "transitioning": false
        },
        "uid": "565234be-5f98-4d0b-a020-ca87f6fcabb6"
      },
      "spec": {
        "devPath": "/dev/sda",
        "fileSystem": {
          "forceFormatted": true,
          "mountPoint": "",
          "provisioned": true
        },
        "nodeName": "dell-poweredge-r720"
      },
      "status": {
        "conditions": [
          {
            "error": false,
            "lastUpdateTime": "2022-09-23T21:42:20Z",
            "message": "Done device ext4 filesystem formatting",
            "status": "False",
            "transitioning": false,
            "type": "Formatting"
          },
          {
            "error": false,
            "lastUpdateTime": "2022-09-23T21:42:21Z",
            "status": "True",
            "transitioning": false,
            "type": "Mounted"
          },
          {
            "error": false,
            "lastUpdateTime": "2022-09-23T21:42:21Z",
            "message": "Added disk dbaa114c4d3cb0f53a83ba607a3a22ef to longhorn node `dell-poweredge-r720` as an additional disk",
            "status": "True",
            "transitioning": false,
            "type": "AddedToNode"
          }
        ],
        "deviceStatus": {
          "capacity": {
            "physicalBlockSizeBytes": 512,
            "sizeBytes": 3000034656256
          },
          "details": {
            "busPath": "pci-0000:02:00.0-scsi-0:2:1:0",
            "deviceType": "disk",
            "driveType": "HDD",
            "model": "PERC_H710",
            "serialNumber": "6c81f660e32a36002ac0e02118e13aef",
            "storageController": "SCSI",
            "uuid": "92b13d02-2c7a-4ffc-a295-852f2f53b5e1",
            "vendor": "DELL",
            "wwn": "0x6c81f660e32a36002ac0e02118e13aef"
          },
          "devPath": "/dev/sdb",
          "fileSystem": {
            "LastFormattedAt": "2022-09-23T21:42:20Z",
            "mountPoint": "/var/lib/harvester/extra-disks/dbaa114c4d3cb0f53a83ba607a3a22ef",
            "type": "ext4"
          },
          "partitioned": false
        },
        "provisionPhase": "Provisioned",
        "state": "Active"
      }
    }
  ]
}
```

And I don't think this is front-end related, because with the: 
GET
	https://192.168.1.166/v1/harvester/longhorn.io.nodes
The Response back is:
```
{
  "type": "collection",
  "links": {
    "self": "https://192.168.1.166/v1/harvester/longhorn.io.nodes"
  },
  "createTypes": {
    "longhorn.io.node": "https://192.168.1.166/v1/harvester/longhorn.io.nodes"
  },
  "actions": {},
  "resourceType": "longhorn.io.node",
  "revision": "1116004",
  "data": [
    {
      "id": "longhorn-system/dell-poweredge-r720",
      "type": "longhorn.io.node",
      "links": {
        "remove": "https://192.168.1.166/v1/harvester/longhorn.io.nodes/longhorn-system/dell-poweredge-r720",
        "self": "https://192.168.1.166/v1/harvester/longhorn.io.nodes/longhorn-system/dell-poweredge-r720",
        "update": "https://192.168.1.166/v1/harvester/longhorn.io.nodes/longhorn-system/dell-poweredge-r720",
        "view": "https://192.168.1.166/apis/longhorn.io/v1beta2/namespaces/longhorn-system/nodes/dell-poweredge-r720"
      },
      "apiVersion": "longhorn.io/v1beta2",
      "kind": "Node",
      "metadata": {
        "creationTimestamp": "2022-09-23T18:05:56Z",
        "fields": [
          "dell-poweredge-r720",
          "True",
          true,
          "True",
          "3d23h"
        ],
        "finalizers": [
          "longhorn.io",
          "wrangler.cattle.io/harvester-ndm-node-handler"
        ],
        "generation": 7,
        "managedFields": [
          {
            "apiVersion": "longhorn.io/v1beta2",
            "fieldsType": "FieldsV1",
            "fieldsV1": {
              "f:spec": {
                "f:disks": {
                  "f:default-disk-452c7f304567d411": {
                    "f:tags": {}
                  }
                },
                "f:tags": {}
              }
            },
            "manager": "harvester",
            "operation": "Update",
            "time": "2022-09-27T00:10:49Z"
          },
          {
            "apiVersion": "longhorn.io/v1beta2",
            "fieldsType": "FieldsV1",
            "fieldsV1": {
              "f:status": {
                "f:diskStatus": {
                  "f:dbaa114c4d3cb0f53a83ba607a3a22ef": {
                    "f:scheduledReplica": {
                      "f:pvc-8f720d6c-3be8-4473-8488-3a752cf8bd4f-r-ace551ca": {}
                    },
                    "f:storageMaximum": {}
                  }
                }
              }
            },
            "manager": "longhorn-manager",
            "operation": "Update",
            "time": "2022-09-27T00:15:22Z"
          },
          {
            "apiVersion": "longhorn.io/v1beta1",
            "fieldsType": "FieldsV1",
            "fieldsV1": {
              "f:metadata": {
                "f:finalizers": {
                  "v:\"wrangler.cattle.io/harvester-ndm-node-handler\"": {}
                }
              },
              "f:spec": {
                "f:disks": {
                  "f:dbaa114c4d3cb0f53a83ba607a3a22ef": {
                    ".": {},
                    "f:allowScheduling": {},
                    "f:evictionRequested": {},
                    "f:path": {},
                    "f:storageReserved": {},
                    "f:tags": {}
                  }
                }
              }
            },
            "manager": "node-disk-manager",
            "operation": "Update",
            "time": "2022-09-27T02:00:04Z"
          },
          {
            "apiVersion": "longhorn.io/v1beta2",
            "fieldsType": "FieldsV1",
            "fieldsV1": {
              "f:status": {
                "f:conditions": {},
                "f:diskStatus": {
                  "f:dbaa114c4d3cb0f53a83ba607a3a22ef": {
                    "f:scheduledReplica": {
                      "f:pvc-bf1114f1-b15d-4678-b51e-aa74f5c9c98e-r-238d7cca": {},
                      "f:pvc-bf1114f1-b15d-4678-b51e-aa74f5c9c98e-r-9adc7f15": {},
                      "f:pvc-bf1114f1-b15d-4678-b51e-aa74f5c9c98e-r-bdcb6c3f": {}
                    },
                    "f:storageAvailable": {},
                    "f:storageScheduled": {}
                  },
                  "f:default-disk-452c7f304567d411": {
                    "f:storageAvailable": {}
                  }
                }
              }
            },
            "manager": "longhorn-manager",
            "operation": "Update",
            "subresource": "status",
            "time": "2022-09-27T16:50:12Z"
          }
        ],
        "name": "dell-poweredge-r720",
        "namespace": "longhorn-system",
        "relationships": [
          {
            "toId": "longhorn-system/instance-manager-e-8c2fc7f1",
            "toType": "longhorn.io.instancemanager",
            "rel": "owner",
            "state": "active",
            "message": "Resource is current"
          },
          {
            "toId": "longhorn-system/instance-manager-r-3abd407d",
            "toType": "longhorn.io.instancemanager",
            "rel": "owner",
            "state": "active",
            "message": "Resource is current"
          }
        ],
        "resourceVersion": "1103523",
        "state": {
          "error": false,
          "message": "Resource is Ready",
          "name": "active",
          "transitioning": false
        },
        "uid": "7bd63fee-f2ff-49b0-ba93-caa73b076a73"
      },
      "spec": {
        "allowScheduling": true,
        "disks": {
          "dbaa114c4d3cb0f53a83ba607a3a22ef": {
            "allowScheduling": true,
            "evictionRequested": false,
            "path": "/var/lib/harvester/extra-disks/dbaa114c4d3cb0f53a83ba607a3a22ef",
            "storageReserved": 0,
            "tags": []
          },
          "default-disk-452c7f304567d411": {
            "allowScheduling": true,
            "evictionRequested": false,
            "path": "/var/lib/harvester/defaultdisk",
            "storageReserved": 255808816742,
            "tags": [
              "ssd"
            ]
          }
        },
        "engineManagerCPURequest": 0,
        "evictionRequested": false,
        "name": "dell-poweredge-r720",
        "replicaManagerCPURequest": 0,
        "tags": [
          "r720"
        ]
      },
      "status": {
        "conditions": [
          {
            "error": false,
            "lastProbeTime": "",
            "lastTransitionTime": "2022-09-23T18:05:56Z",
            "lastUpdateTime": "2022-09-23T18:05:56Z",
            "message": "",
            "reason": "",
            "status": "True",
            "transitioning": false,
            "type": "MountPropagation"
          },
          {
            "error": false,
            "lastProbeTime": "",
            "lastTransitionTime": "2022-09-27T16:50:12Z",
            "lastUpdateTime": "2022-09-27T16:50:12Z",
            "message": "Node dell-poweredge-r720 is ready",
            "reason": "",
            "status": "True",
            "transitioning": false,
            "type": "Ready"
          },
          {
            "error": false,
            "lastProbeTime": "",
            "lastTransitionTime": "2022-09-23T18:05:56Z",
            "lastUpdateTime": "2022-09-23T18:05:56Z",
            "message": "",
            "reason": "",
            "status": "True",
            "transitioning": false,
            "type": "Schedulable"
          }
        ],
        "diskStatus": {
          "dbaa114c4d3cb0f53a83ba607a3a22ef": {
            "conditions": [
              {
                "lastProbeTime": "",
                "lastTransitionTime": "2022-09-24T03:45:18Z",
                "message": "Disk dbaa114c4d3cb0f53a83ba607a3a22ef(/var/lib/harvester/extra-disks/dbaa114c4d3cb0f53a83ba607a3a22ef) on node dell-poweredge-r720 is schedulable",
                "reason": "",
                "status": "True",
                "type": "Schedulable"
              },
              {
                "lastProbeTime": "",
                "lastTransitionTime": "2022-09-24T03:45:18Z",
                "message": "Disk dbaa114c4d3cb0f53a83ba607a3a22ef(/var/lib/harvester/extra-disks/dbaa114c4d3cb0f53a83ba607a3a22ef) on node dell-poweredge-r720 is ready",
                "reason": "",
                "status": "True",
                "type": "Ready"
              }
            ],
            "diskUUID": "a0bea76a-ba61-4485-a95d-fe461932c636",
            "scheduledReplica": {
              "pvc-8f720d6c-3be8-4473-8488-3a752cf8bd4f-r-ace551ca": 375809638400,
              "pvc-bf1114f1-b15d-4678-b51e-aa74f5c9c98e-r-238d7cca": 64424509440,
              "pvc-bf1114f1-b15d-4678-b51e-aa74f5c9c98e-r-9adc7f15": 64424509440,
              "pvc-bf1114f1-b15d-4678-b51e-aa74f5c9c98e-r-bdcb6c3f": 64424509440
            },
            "storageAvailable": 2928777625600,
            "storageMaximum": 2951778361344,
            "storageScheduled": 569083166720
          },
          "default-disk-452c7f304567d411": {
            "conditions": [
              {
                "lastProbeTime": "",
                "lastTransitionTime": "2022-09-23T18:05:56Z",
                "message": "Disk default-disk-452c7f304567d411(/var/lib/harvester/defaultdisk) on node dell-poweredge-r720 is ready",
                "reason": "",
                "status": "True",
                "type": "Ready"
              },
              {
                "lastProbeTime": "",
                "lastTransitionTime": "2022-09-23T18:05:56Z",
                "message": "Disk default-disk-452c7f304567d411(/var/lib/harvester/defaultdisk) on node dell-poweredge-r720 is schedulable",
                "reason": "",
                "status": "True",
                "type": "Schedulable"
              }
            ],
            "diskUUID": "aace2402-b1e1-4dcc-8be4-c0fc673f1060",
            "scheduledReplica": {
              "pvc-5d5661a6-ba59-4166-bad0-fe241075ed18-r-7b31154f": 53687091200,
              "pvc-78e7cb37-2c69-4c13-bc33-367025da2de9-r-d2850e37": 2147483648,
              "pvc-e50f0b2e-dc05-4b17-b86e-4fe59e930927-r-ea147d5f": 5368709120
            },
            "storageAvailable": 849346560000,
            "storageMaximum": 852696055808,
            "storageScheduled": 61203283968
          }
        },
        "region": "",
        "zone": ""
      }
    }
  ]
}
```

From that the "extra disk" that's used, "tags" is an empty list/array:
```
        "disks": {
          "dbaa114c4d3cb0f53a83ba607a3a22ef": {
            "allowScheduling": true,
            "evictionRequested": false,
            "path": "/var/lib/harvester/extra-disks/dbaa114c4d3cb0f53a83ba607a3a22ef",
            "storageReserved": 0,
            "tags": []
          },
```

Is there a way to collect logs from Longhorn when disk tags get created and validate those?
I powered off my R720 at 23:55 on 09/26/22 - powered it back on at 09:00 on 09/27/22.
So roughly 11hrs powered off, on the single node.
I could see someone needing to take a node offline for that amount of time possibly in order to change/upgrade/replace hardware.
Then when the node comes back online the disk tags would be missing.

Additional support bundle:
[supportbundle_22241fc2-4f6e-4105-9483-178b97ed77b8_2022-09-27T17-40-25Z.zip](https://github.com/harvester/harvester/files/9658472/supportbundle_22241fc2-4f6e-4105-9483-178b97ed77b8_2022-09-27T17-40-25Z.zip)

Perhaps a way to reproduce would be:
- set up extra disk, format in harvester
- add tags to the disk
- create storage class that would target the disk
- poweroff node for 8hrs
- power back on node, tags should still be present on the disk

But I'm still not sure if that way to reproduce would work.

**Comment by Vicente-Cheng** [comment]
**QUICK Reproduce method: delete the _node-disk-manager_ pod**

**Comment by Vicente-Cheng** [comment]
Another reproduce method:
https://github.com/harvester/node-disk-manager/pull/42#issuecomment-1261129222

**Comment by irishgordo** [comment]
@Vicente-Cheng @futuretea @bk201 
Thanks for your work on this!

I validated on: `master-75876dee-head` ISO w/ qemu/kvm
This all looks good!

I removed the `node-disk-manager` pod several times and also ran through a series of restarts of the node to make sure the labels were still present on the disks.  Both in Harvester UI & Longhorn UI.
 
![Screenshot from 2022-10-04 10-21-26](https://user-images.githubusercontent.com/5370752/193885986-6f2e5d98-d827-4a20-9007-fbf2dc3c5fc2.png)
![Screenshot from 2022-10-04 10-06-41](https://user-images.githubusercontent.com/5370752/193885989-60769a24-1e58-470c-bdd7-d0e73f81934f.png)

Thanks again!
I'll go ahead and close this out :smile:
